### ISSUE TITLE
A strict version of Extract

### ISSUE THREAD

**Issue by Vages** [sindresorhus/type-fest#222]
I have been looking for a strict version of `Extract<T, U>`. For simplicity, let's call it `ExtractStrict<T, U>`. `ExtractStrict` would have one additional feature compared to `Extract`: `ExtractStrict` would throw an error if some element in `U` is not part of `T`.

I know of an implementation by @ChristianIvicevic which has been proposed for inclusion in Typescript itself: https://github.com/microsoft/TypeScript/issues/31474 I've asked him if he'd be willing to submit it here instead.

**Comment by sindresorhus** [comment]
Sounds useful. A pull request would be welcome.

**Comment by Vages** [comment]
I've found another use case in my code for a strict version of Exclude too, so I've added it to the title.

**Comment by Vages** [comment]
I'll wait until next Thursday for an answer from @ChristianIvicevic; he deserves the chance to  get some credit for his contribution.

Otherwise, I'll write my own implementation and submit a PR.

**Comment by ChristianIvicevic** [comment]
Hey folks, I'll be taking a look at it this week. Quick question about tests @sindresorhus: I noticed that the tests are used to assert what resulting types look like. Does the project use anything to handle certain test cases that are expected to yield invalid results in order to test erroneous type arguments?

**Comment by sindresorhus** [comment]
Yes, there's an `expectError` assertion: https://github.com/SamVerschueren/tsd#expecterrorfunction

**Comment by Vages** [comment]
Here's my implementation of ExcludeStrict, @ChristianIvicevic (a derivative of your ExtractStrict): `export type ExcludeStrict<T, U extends T> = Exclude<T, U>;`.

**Comment by younho9** [comment]
> Here's my implementation of ExcludeStrict, @ChristianIvicevic (a derivative of your ExtractStrict): `export type ExcludeStrict<T, U extends T> = Exclude<T, U>;`.

Could ExtractStrict be implemented in the same way? `export type ExtractStrict<T, U extends T> = Extract<T, U>;`

**Comment by Vages** [comment]
I'll put up a PR for this in a few hours.

**Comment by NN---** [comment]
You can use ts-essentials package which has these definitions.

https://github.com/ts-essentials/ts-essentials#StrictExtract

**Comment by Vages** [comment]
Thank you for the tip, @NN--- . In that case, I don't see any point in fixing PR #291 :/

**Comment by sindresorhus** [comment]
If anyone wants to work on this, see the initial attempt in https://github.com/sindresorhus/type-fest/pull/291.

**Comment by nigeltroy** [comment]
Hey y'all, I know that this issue is pretty stale, but I have implemented `ExtractStrict` in [this PR](https://github.com/sindresorhus/type-fest/pull/1119) (will do `ExcludeStrict` next), but I did change some functionality that was provided here: for `ExtractStrict<T, U>`, instead of requiring `U extends T`, I now have `U extends Partial<Exact<T, U>>`. This accomplishes the following:

1. Still validates that the fields in U are present in T (like `U extends T` does)
2. You don't have to provide the ENTIRE subtype to extract a union member, you just have to provide a partial of the subtype that `Extract` can use to discriminate.

I have added examples and test cases in the PR. Let me know what you think!

**Comment by Perdolique** [comment]
@sindresorhus @Vages How to use it from this package? It doesn't look exported like other helpers. Not sure it's a bug.

**Comment by som-sm** [comment]
@Perdolique These types aren’t available yet, they’ll be available with our next major release.
